### DISCUSSION TITLE
Count keys with same prefix in one query

### DISCUSSION THREAD

**Discussion by rei-ber** [datalust/seq-tickets#1501] [Help] [answered]
Hi people :)

For example I have the following two data objects in seq: 

```json
{
"key_1": "value_1",
"key_2": "value_2"
}
{
"key_1": "value_3",
"key_2": "value_1",
"key_3": "value_2"
}
```

The aim now is, to count and order the key-value-pairs in one query. The keys have the same prefix `key_` followed by an number. Some keys may not exists, so I have to filter null keys from the result. The result should be as follows:

```
"value_1": 2,
"value_2": 2,
"value_3": 1
```

The seq query language is not as much powerful as sql is, I guess. I would resolve it with with nested selects, but it is not possible. I already played a lot with the supported functions, aggregations and other possibilities of seqs' query language, but nothing seems to help. The only query I got to work for one key is:

```
select count(key_1) as count
from stream 
group by key_1
order by count desc
limit 10
```

Any ideas or is it not possible at all?

**Comment by nblumhardt** [comment, marked as answer]
Great question - I think you're right, in that there is no single query capable of providing this, today.

The following works to some degree:

```
select distinct([key_1, key_2, key_3]) from stream
```

producing:

```
null
value_3
value_1
value_2
```

but `distinct` over arrays is special-cased, and adding `count()` as below doesn't work:

```
select count(distinct([key_1, key_2, key_3])) from stream
```

Using a query like:

```
select key_1, key_2, key_3
from stream
```

and exporting to CSV would open up options like post-processing with Excel or another programming language - the best I can currently come up with :-)

More thought going into this area 👍

**Reply by rei-ber** [reply]
Thanks for your answer :) 

> select count(distinct([key_1, key_2, key_3])) from stream

This works for me, but does not produce the result I want ;) even though I append a `group by` and `order by`.

The main aim is to collect all equal keys together and count them. The other problem then is, that the `group by` does not allow grouping each key separately. Naturally it combines each key with the others and counts these combinations. 

I was already thinking, to modify the data structure. Like, putting all keys in a list into the data, for example. But it would count all lists with the same content and not each element in the lists separately.

> and exporting to CSV would open up options like post-processing with Excel or another programming language - the best I can currently come up with :-)

That's true, it opens much more options. But exporting to CSV is nothing I would like to do. From the query result, I would like to create a pie chart in seq directly.

> More thought going into this area +1

That would be nice; thanks for your help again :)
